_source/_posts/2022-02-28-spring-native-jhipster.adoc

```diff
@@ -0,0 +1,451 @@
+- Allowed Logout URLs: `\http://localhost:8080/`
+
+Scroll to the bottom and click **Save Changes**.
+
+Copy your Auth0 domain, client ID, and client secret into the `.auth0.env` file you created earlier. Then, run `source .auth0.env`.
+
+In the https://manage.auth0.com/#/roles[roles] section, create new roles named `ROLE_ADMIN` and `ROLE_USER`.
+
+Create a new user account in the https://manage.auth0.com/#/users[users] section. Click on the **Role** tab to assign the roles you just created to the new account.
+
+_Make sure your new user's email is verified before attempting to log in!_
+
+Next, head to **Auth Pipeline** > **Rules** > **Create**. Select the `Empty rule` template. Provide a meaningful name like `Group claims` and replace the Script content with the following.
+
+[source,js]
+----
+function(user, context, callback) {
+  user.preferred_username = user.email;
+  const roles = (context.authorization || {}).roles;
+
+  function prepareCustomClaimKey(claim) {
+    return `https://www.jhipster.tech/${claim}`;
+  }
+
+  const rolesClaim = prepareCustomClaimKey('roles');
+
+  if (context.idToken) {
+    context.idToken[rolesClaim] = roles;
+  }
+
+  if (context.accessToken) {
+    context.accessToken[rolesClaim] = roles;
+  }
+
+  callback(null, user, context);
+}
+----
+
+This code is adding the user's roles to a custom claim (prefixed with `https://www.jhipster.tech/roles`). Click **Save changes** to continue.
+
+**NOTE**: Want to have all these steps automated for you? Vote for https://github.com/auth0/auth0-cli/issues/351[issue #351] in the Auth0 CLI project.
+
+=== Run your native JHipster app
+
+After you've built your app, it will be available in `target/native-executable`. Start Keycloak or source your Okta/Auth0 settings. Then, run `./target/native-executable`. It should start in under around one second!
+
+++++
+{% twitter 1494118802988089348 %}
+++++
+
+// todo: I like showing the tweet ^^ because it got so many likes, but maybe just the screenshot would be better?
+
+== What does the JHipster Native blueprint do?
+
+The JHipster Native plugin integrates Spring Native into a JHipster project based on https://github.com/mraible/spring-native-examples#readme[the findings from Josh Long and my research]. I documented our findings in September and December 2021.
+
+- Sep 30, 2021: https://www.linkedin.com/pulse/jhipster-works-spring-native-matt-raible/[JHipster Works with Spring Native!]
+- Dec 14, 2021: https://www.linkedin.com/pulse/jhipster-works-spring-native-part-2-matt-raible/[JHipster works with Spring Native, Part 2!]
+
+The crazy thing is, one of the hardest problems we had to solve was with JPA and relationships. In JVM mode, everything worked fine. When running in native mode, https://github.com/jhipster/generator-jhipster/issues/17794[there was an exception]. The solution took days to figure out, but was quite simple to fix: I just had to add a type hint for `java.util.HashSet.class`. 🤯
```

Comment:
```suggestion
The crazy thing is, one of the hardest problems we had to solve was with JPA and relationships. In JVM mode, everything worked fine. When running in native mode, https://github.com/jhipster/generator-jhipster/issues/17794[there was an exception]. The solution took days to figure out, but was quite simple to fix: I just had to add a type hint for `java.util.HashSet.class`. 🤯
```
I'm quite fond of the word "crazy" but I'm teaching myself not to use it because it is viewed as offensive : 

https://ncdj.org/2016/09/terms-to-avoid-when-writing-about-disability-2/. 
"Crazy, insane, nuts, psycho: All are considered offensive and should not be used except in direct quotes. See mental health. "

Reply:
I'm so used to saying "the crazy thing is". It doesn't sound quite right with the other suggestions, but I'll use "unexpected". Or maybe "the wild thing is" would work here?